roachtest: npgsql failed

roachtest.npgsql [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12562959?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12562959?buildTab=artifacts#/npgsql) on release-23.1 @ [a4e0bfde6a520c605276ed73fdcc9bb3db20949c](https://github.com/cockroachdb/cockroach/commits/a4e0bfde6a520c605276ed73fdcc9bb3db20949c):


```
(orm_helpers.go:212).summarizeFailed: 
Tests run on Cockroach v23.1.12-dev-a4e0bfde6a520c605276ed73fdcc9bb3db20949c
Tests run against npgsql v7.0.2
3278 Total Tests Run
2571 tests passed
707 tests failed
248 tests skipped
63 tests ignored
0 tests passed unexpectedly
1 test failed unexpectedly
0 tests expected failed but skipped
0 tests expected failed but not run
---
--- FAIL: Npgsql.Tests.CommandTests(Multiplexing).Input_and_output_parameters(Default) -  (unexpected)
For a full summary look at the npgsql artifacts 
An updated blocklist (npgsqlBlocklist) is available in the artifacts' npgsql log
test artifacts and logs in: /artifacts/npgsql/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-12562959/npgsql/1699458912129/1699459271047)

</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*npgsql.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33310

--- a/pkg/cmd/roachtest/tests/npgsql.go
+++ b/pkg/cmd/roachtest/tests/npgsql.go
@@ -78,7 +78,7 @@ func registerNpgsql(r registry.Registry) {
 			c,
 			node,
 			"install dependencies",
-			`sudo snap install dotnet-sdk --classic && \
+			`sudo snap install dotnet-sdk --channel=7.0/stable --classic && \
 sudo snap alias dotnet-sdk.dotnet dotnet && \
 sudo ln -s /snap/dotnet-sdk/current/dotnet /usr/local/bin/dotnet`,
 		); err != nil {

--- a/pkg/sql/logictest/testdata/logic_test/read_committed
+++ b/pkg/sql/logictest/testdata/logic_test/read_committed
@@ -424,3 +424,45 @@ subtest end
 
 statement ok
 SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL SERIALIZABLE
+
+subtest cursor
+
+# Verify that SQL and PL/pgSQL cursors run under READ COMMITTED correctly.
+statement ok
+BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
+DECLARE curs CURSOR FOR SELECT * FROM (VALUES (1), (2), (3)) AS t;
+MOVE FORWARD 2 IN curs;
+
+query I
+FETCH 1 curs
+----
+3
+
+statement ok
+COMMIT
+
+statement ok
+CREATE FUNCTION f_cursor() RETURNS INT AS $$
+DECLARE
+  foo INT;
+  curs CURSOR FOR VALUES (1), (2);
+BEGIN
+  OPEN curs;
+  FETCH curs INTO foo;
+  CLOSE curs;
+  RETURN foo;
+END
+$$ LANGUAGE plpgsql
+
+statement ok
+BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED
+
+query I
+SELECT f_cursor()
+----
+1
+
+statement ok
+COMMIT
+
+subtest end

--- a/pkg/sql/sql_cursor.go
+++ b/pkg/sql/sql_cursor.go
@@ -59,7 +59,17 @@ func (p *planner) DeclareCursor(ctx context.Context, s *tree.DeclareCursor) (pla
 				sd = sd.Clone()
 				sd.StmtTimeout = 0
 			}
-			ie := p.ExecCfg().InternalDB.NewInternalExecutor(sd)
+			// We avoid using the internal executor provided by p.InternalSQLTxn()
+			// since we want to customize the session data used by the cursor.
+			ief := p.ExecCfg().InternalDB
+			ie := MakeInternalExecutor(ief.server, ief.memMetrics, ief.monitor)
+			ie.SetSessionData(sd)
+			ie.extraTxnState = &extraTxnState{
+				txn:                p.Txn(),
+				descCollection:     p.Descriptors(),
+				jobs:               p.extendedEvalCtx.jobs,
+				schemaChangerState: p.extendedEvalCtx.SchemaChangerState,
+			}
 			if err := p.checkIfCursorExists(s.Name); err != nil {
 				return nil, err
 			}

